Userland/Utilities/traceroute.cpp

```diff
@@ -0,0 +1,175 @@
+
+    if (max_retries < 1) {
+        warnln("Invalid maximum retries amount");
+        return 1;
+    }
+
+    auto* hostent = gethostbyname(host_name);
+    if (!hostent) {
+        warnln("Lookup failed for '{}'", host_name);
+        return 1;
+    }
+    sockaddr_in host_address {};
+    memset(&host_address, 0, sizeof(host_address));
+    host_address.sin_family = AF_INET;
+    host_address.sin_port = 44444;
+    host_address.sin_addr.s_addr = *(const in_addr_t*)hostent->h_addr_list[0];
+
+    int fd = socket(AF_INET, SOCK_RAW, IPPROTO_ICMP);
+    if (fd < 0) {
+        perror("socket");
+        return 1;
+    }
+
+    if (setgid(getgid()) || setuid(getuid())) {
+        warnln("Failed to drop privileges");
+        return 1;
+    }
+
+    if (pledge("stdio inet unix", nullptr) < 0) {
+        perror("pledge");
+        return 1;
+    }
+
+    timeval timeout { 5, 0 };
+    if (setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout)) < 0) {
+        perror("setsockopt");
+        return 1;
+    }
+
+    struct icmp_request {
+        struct icmphdr header;
+        char msg[64 - sizeof(struct icmphdr)];
+    };
+
+    struct icmp_response {
+        u8 ip_header[20];
+        struct icmphdr header;
+        char msg[64 - sizeof(struct icmphdr)];
+    };
+
+    // 1: reached target
+    // 0: got ttl exhausted response
+    // -1: error or no response
+    auto try_reach_host = [&](int ttl) -> int {
+        Core::ElapsedTimer m_timer { true };
+        for (auto i = 0; i < max_retries; i++) {
+            icmp_request request {};
+            request.header = { ICMP_ECHO, 0, 0, { { 0, 0 } } };
+            VERIFY(String::number(ttl).copy_characters_to_buffer(request.msg, sizeof(request.msg)));
+            request.header.checksum = internet_checksum(&request, sizeof(request));
+
+            m_timer.start();
+            if (sendto(fd, &request, sizeof(request), 0, (sockaddr*)&host_address, sizeof(host_address)) < 0)
+                return -1;
+
+            icmp_response response {};
+            sockaddr_in peer_address {};
+            memset(&peer_address, 0, sizeof(peer_address));
+            size_t peer_address_size = sizeof(peer_address);
+            int result = recvfrom(fd, &response, sizeof(response), 0, (sockaddr*)&peer_address, (socklen_t*)&peer_address_size);
+            if (result < 0) {
+                if (result == EAGAIN)
+                    return -1;
+                continue;
+            }
+
+            if (response.header.type != ICMP_ECHOREPLY && response.header.type != ICMP_TIME_EXCEEDED)
+                continue;
+
+            auto* peer = gethostbyaddr(&peer_address.sin_addr, sizeof(peer_address.sin_addr), AF_INET);
+
+            const char* peer_name;
+            if (peer) {
+                peer_name = peer->h_name;
+            } else {
+                char addr_buf[INET_ADDRSTRLEN];
+                peer_name = inet_ntop(AF_INET, &peer_address.sin_addr, addr_buf, sizeof(addr_buf));
```

Comment:
> The inet_ntop() function shall return a pointer to the buffer containing the text string if the conversion succeeds, and NULL otherwise, and set errno to indicate the error.

The else branch here returns a pointer into the extra-scoped addr_buf whose lifetime ends at the end of the else statement. I suggest moving the declaration of that buffer up next to peer_name, or making peer_name an AK::String.

Reply:
good catch, fixed :)